### PR TITLE
Make ReplaceTypeParametersBasedOnTypeConstraints asynchronous

### PR DESCRIPTION
**Pull request by sharwell**
Fixes an exception caused by calling `WaitAndGetResult` on a background thread. Rather than use a blocking wait, the entire stack is made asynchronous.

### CODE REVIEW

File: src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.ReplaceTypeParameterBasedOnTypeConstraintVisitor.cs [hunk 1]
```diff
@@ -98,13 +107,13 @@ public override ITypeSymbol VisitTypeParameter(ITypeParameterSymbol symbol)
                         if (symbol.ConstraintTypes.All(t => t is INamedTypeSymbol))
                         {
                             var immutableProjects = _solution.Projects.ToImmutableHashSet();
-                            var derivedImplementedTypesOfEachConstraintType = symbol.ConstraintTypes.Select(ct =>
+                            var derivedImplementedTypesOfEachConstraintType = (await Task.WhenAll(symbol.ConstraintTypes.Select(async ct =>
```
Comment (tmat) [inline src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.ReplaceTypeParameterBasedOnTypeConstraintVisitor.cs]: Alternative approach would be to push the async operation even higher up - store `symbol.ConstraintTypes` here to some result builder list and and perform the processing that's in the lambda outside of the visitor - the visitor would return the builder and then the caller would run the SymbolFinder on the symbols stored in the builder.

Comment (sharwell) [reply src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.ReplaceTypeParameterBasedOnTypeConstraintVisitor.cs]: This seems like a lot of work to refactor the behavior, and I'm not sure the result would be as easy to read. Is there a particular advantage you'd be hoping to realize from such an effort?

Comment (tmat) [reply src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.ReplaceTypeParameterBasedOnTypeConstraintVisitor.cs]: IDK how much work that is, so maybe not. But in general I think it seems cleaner to separate the two steps - one step is analyze the symbols (visitor) and then act on the results in a separate step.

File: src/Features/Core/Portable/GenerateMember/GenerateParameterizedMember/AbstractGenerateParameterizedMemberService.SignatureInfo.cs [hunk 2]
```diff
@@ -196,10 +197,10 @@ public ITypeSymbol DetermineReturnType(CancellationToken cancellationToken)
                     : ImmutableArray.Create(throwStatement);
             }
 
-            private ImmutableArray<IParameterSymbol> DetermineParameters(CancellationToken cancellationToken)
+            private async ValueTask<ImmutableArray<IParameterSymbol>> DetermineParametersAsync(CancellationToken cancellationToken)
             {
                 var modifiers = DetermineParameterModifiers(cancellationToken);
-                var types = DetermineParameterTypes(cancellationToken).Select(t => FixType(t, cancellationToken)).ToList();
+                var types = (await Task.WhenAll(DetermineParameterTypes(cancellationToken).Select(t => FixTypeAsync(t, cancellationToken).AsTask())).ConfigureAwait(false)).ToList();
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/GenerateMember/GenerateParameterizedMember/AbstractGenerateParameterizedMemberService.SignatureInfo.cs]: nit, extract local to make this more readable.

File: src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.ReplaceTypeParameterBasedOnTypeConstraintVisitor.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.ReplaceTypeParameterBasedOnTypeConstraintVisitor.cs]: i'm ok with teh new approach myself.

File: src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.ReplaceTypeParameterBasedOnTypeConstraintVisitor.cs [hunk 3]
```diff
@@ -51,9 +49,9 @@ public override ITypeSymbol VisitArrayType(IArrayTypeSymbol symbol)
                 return _compilation.CreateArrayTypeSymbol(elementType, symbol.Rank);
             }
 
-            public override ITypeSymbol VisitNamedType(INamedTypeSymbol symbol)
+            public override async ValueTask<ITypeSymbol> VisitNamedType(INamedTypeSymbol symbol)
             {
-                var arguments = symbol.TypeArguments.Select(t => t.Accept(this)).ToArray();
+                var arguments = (await Task.WhenAll(symbol.TypeArguments.Select(t => t.Accept(this).AsTask())).ConfigureAwait(false)).ToArray();
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.ReplaceTypeParameterBasedOnTypeConstraintVisitor.cs]: this pattern seems to appear a lot.  can we extract out a helper to simplify this. i.e. `await RecurseIntArray(symbol.TypeArguments).ConfigureAwait(false)`?

File: src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.ReplaceTypeParameterBasedOnTypeConstraintVisitor.cs [hunk 4]
```diff
@@ -98,13 +96,13 @@ public override ITypeSymbol VisitTypeParameter(ITypeParameterSymbol symbol)
...
                                 var derivedAndImplementedTypes = new List<INamedTypeSymbol>();
-                                var derivedClasses = SymbolFinder.FindDerivedClassesAsync((INamedTypeSymbol)ct, _solution, immutableProjects, _cancellationToken).WaitAndGetResult(_cancellationToken);
-                                var implementedTypes = DependentTypeFinder.FindTransitivelyImplementingStructuresAndClassesAsync((INamedTypeSymbol)ct, _solution, immutableProjects, _cancellationToken).WaitAndGetResult(_cancellationToken);
+                                var derivedClasses = await SymbolFinder.FindDerivedClassesAsync((INamedTypeSymbol)ct, _solution, immutableProjects, _cancellationToken).ConfigureAwait(false);
+                                var implementedTypes = await DependentTypeFinder.FindTransitivelyImplementingStructuresAndClassesAsync((INamedTypeSymbol)ct, _solution, immutableProjects, _cancellationToken).ConfigureAwait(false);
                                 return derivedClasses.Concat(implementedTypes.Select(s => s.Symbol)).ToList();
-                            }).ToList();
+                            })).ConfigureAwait(false)).ToList();
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.ReplaceTypeParameterBasedOnTypeConstraintVisitor.cs]: would def prefer to introduce locals for `await` calls.  the wrapping with parens and continuing operations is not trivial to read.

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Utilities/AsyncSymbolVisitor.cs [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading.Tasks;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal abstract class AsyncSymbolVisitor : SymbolVisitor<ValueTask>
+    {
+    }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Utilities/AsyncSymbolVisitor.cs]: is this type valuable enough to have?

File: src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.ReplaceTypeParameterBasedOnTypeConstraintVisitor.cs (hunk 3 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.ReplaceTypeParameterBasedOnTypeConstraintVisitor.cs]: Extracted `SpecializedTasks.WhenAll` to operate on `ValueTask<T>`.

File: src/Features/Core/Portable/GenerateMember/GenerateParameterizedMember/AbstractGenerateParameterizedMemberService.SignatureInfo.cs (hunk 2 above)
Comment (sharwell) [reply src/Features/Core/Portable/GenerateMember/GenerateParameterizedMember/AbstractGenerateParameterizedMemberService.SignatureInfo.cs]: Now simplified

File: src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.ReplaceTypeParameterBasedOnTypeConstraintVisitor.cs (hunk 4 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.ReplaceTypeParameterBasedOnTypeConstraintVisitor.cs]: This should be simpler now (no operations changed on `await` expressions)

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Utilities/AsyncSymbolVisitor.cs (hunk 5 above)
Comment (sharwell) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Utilities/AsyncSymbolVisitor.cs]: Not sure. It seemed weird to have one but not the other.

**Pull request merged**